rollup/core/classes/RollupState.cls

```diff
@@ -65,8 +116,286 @@ public without sharing virtual class RollupState {
+      };
     }
   }
+
+  /**
+   * In theory, GenericInfo is overkill for what it does - any rollup operation that can be reduced to a single value
+   * COULD simply take whatever the current value on the parent record is as the source of truth. That being said, the
+   * prior implementation with `Database.Stateful` shamefully split the logic between `RollupAsyncProcessor` and `RollupState`
+   * when managing prior values, and duplicating _some_ data here (as far as what eventually gets persisted to the database as `RollupState__c` records)
+   * seems vastly preferable as opposed to the logic living in two different places
+   */
+  public virtual class GenericInfo extends RollupState {
+    public Object value;
+
+    public void setValue(Object newValue) {
+      this.value = newValue;
+    }
+
+    public virtual override Boolean isEmpty() {
+      return this.value == null;
+    }
+
+    public virtual override Map<String, Object> getUntypedState() {
+      return new Map<String, Object>{ 'key' => this.key, 'keyLength' => this.keyLength, 'typeName' => GenericInfo.class.getName(), 'value' => this.value };
+    }
+  }
+
+  public void loadState(String jobId, Set<String> relatedRecordKeys) {
+    this.jobIds.add(jobId);
+    List<RollupState__c> matchingState = this.loadOrRetrieveCachedState(jobId, relatedRecordKeys);
+    for (RollupState__c state : matchingState) {
+      if (this.statefulPreviouslyRetrievedStateIds.contains(state.Id) == false && state.Body0__c != null) {
+        List<Object> localUncastStates = (List<Object>) new DataWeaveScriptResource.jsonToRollupState()
+          .execute(new Map<String, Object>{ 'records' => '[' + state.Body0__c + ']' })
+          .getValue();
+        for (Object uncastState : localUncastStates) {
+          RollupState castState = (RollupState) uncastState;
+          KEY_TO_STATE.put(castState.key, castState);
+        }
+      }
+      this.statefulPreviouslyRetrievedStateIds.add(state.Id);
+    }
+  }
+
+  public void commitState(String jobId) {
+    this.jobIds.add(jobId);
+    this.populateRelatedRecordStates(jobId);
+  }
+
+  public String cleanup(Set<String> jobIds) {
+    this.jobIds.addAll(jobIds);
+    RollupLogger.Instance.log('about to clean up rollup state', this.jobIds, System.LoggingLevel.INFO);
+    return System.enqueueJob(this);
+  }
+
+  @SuppressWarnings('PMD.ApexCrudViolation')
+  public void execute(System.QueueableContext qc) {
+    System.attachFinalizer(this);
+    List<RollupState__c> matchingState = [
+      SELECT Id
+      FROM RollupState__c
+      WHERE RelatedJobId__c = :this.jobIds
+      LIMIT :(Limits.getLimitDmlRows() - (Limits.getDmlRows() + 1))
+    ];
+    Database.delete(matchingState, false, System.AccessLevel.SYSTEM_MODE);
+  }
+
+  public void execute(System.FinalizerContext fc) {
+    if ([SELECT COUNT() FROM RollupState__c WHERE Id = :this.jobIds LIMIT 1] > 0) {
+      new RollupState().cleanup(new Set<String>(this.jobIds));
```

Comment:
`this.jobIds` is already an instance of `Set<String>` - do you need the extra `new Set` call here? The `cleanup` method then just adds the passed set to _another_ set, so offhand I can't see a need for `new Set<String>()`

```suggestion
      new RollupState().cleanup(this.jobIds);
```

Reply:
Good call, yes!